frappe/oauth.py

```diff
@@ -164,28 +154,72 @@ def validate_code(self, client_id, code, client, request, *args, **kwargs):
 		# Validate the code belongs to the client. Add associated scopes,
 		# state and user to request.scopes and request.user.
 
-		validcodes = frappe.get_all("OAuth Authorization Code", filters={"client": client_id, "validity": "Valid"})
+		validcodes = frappe.get_all(
+			"OAuth Authorization Code",
+			filters={"client": client_id, "validity": "Valid"},
+		)
 
 		checkcodes = []
 		for vcode in validcodes:
 			checkcodes.append(vcode["name"])
 
 		if code in checkcodes:
-			request.scopes = frappe.db.get_value("OAuth Authorization Code", code, 'scopes').split(get_url_delimiter())
-			request.user = frappe.db.get_value("OAuth Authorization Code", code, 'user')
+			request.scopes = frappe.db.get_value(
+				"OAuth Authorization Code", code, "scopes"
+			).split(get_url_delimiter())
+			request.user = frappe.db.get_value("OAuth Authorization Code", code, "user")
+			code_challenge_method = frappe.db.get_value(
+				"OAuth Authorization Code", code, "code_challenge_method"
+			)
+			code_challenge = frappe.db.get_value(
+				"OAuth Authorization Code", code, "code_challenge"
+			)
+
+			if code_challenge and not request.code_verifier:
+				if frappe.db.exists("OAuth Authorization Code", code):
+					frappe.delete_doc(
+						"OAuth Authorization Code", code, ignore_permissions=True
+					)
```

Comment:
Why are we deleting the auth code?

Reply:
`if code_challenge and not request.code_verifier` means the client app that generated auth request had mentioned code_challenge and expects this security. If the app trying to make second call to exchange code with token may be trying to bypass code_challenge, so just deleting the code immediately and invalidating the retries if any.

Reply:
Ah gotcha, that makes sense